compiler/src/iree/compiler/ExternalInterfaces/EncodingExternalModels.cpp

```diff
@@ -0,0 +1,134 @@
+  bool isPropagable(Attribute attr, Value target) const {
+    auto encoding = cast<IREE::Encoding::MatmulKAttr>(attr);
+    Operation *attachedToOperation = target.getDefiningOp();
+    if (!attachedToOperation) {
+      return false;
+    }
+    return TypeSwitch<Operation *, bool>(attachedToOperation)
+        .Case<tensor::CollapseShapeOp>([&](auto op) {
+          auto collapseOp = cast<tensor::CollapseShapeOp>(op);
+          ArrayRef<int32_t> kDims = encoding.getKDims().asArrayRef();
+          // TODO: Relax the check to allow transforming innermost reduction
+          // dimensions. We need to revisit the matmul_k encoding semantic.
+          SmallVector<ReassociationIndices, 4> reassociationMaps =
+              collapseOp.getReassociationIndices();
+          for (int32_t k : kDims) {
+            if (reassociationMaps[k].size() != 1) {
+              return false;
+            }
+          }
+          return true;
+        })
+        .Default([&](auto) { return false; });
+  }
+
+  FailureOr<IREE::Encoding::PropagationEncoding>
+  generateEncodings(Attribute attr, Value target) const {
+    auto encoding = cast<IREE::Encoding::MatmulKAttr>(attr);
+    return TypeSwitch<Operation *,
+                      FailureOr<IREE::Encoding::PropagationEncoding>>(
+               target.getDefiningOp())
+        .Case<tensor::CollapseShapeOp>([&](auto op) {
+          auto collapseOp = cast<tensor::CollapseShapeOp>(op);
+          ArrayRef<int32_t> kDims = encoding.getKDims().asArrayRef();
+          SmallVector<ReassociationIndices, 4> reassociationMaps =
+              collapseOp.getReassociationIndices();
+          // Get a mapping from original iteration space to expanded iteration
+          // space.
+          SmallVector<int32_t> newKDims;
+          for (int32_t kDim : kDims) {
+            newKDims.append(reassociationMaps[kDim].begin(),
+                            reassociationMaps[kDim].end());
+          }
+          MLIRContext *ctx = collapseOp.getContext();
+          auto resultEncodingAttr =
+              IREE::Encoding::MatmulKAttr::get(ctx, newKDims);
+
+          IREE::Encoding::PropagationEncoding propEncoding;
+          propEncoding.operandEncodings.push_back(resultEncodingAttr);
+          // The result encoding will be the same as the encoding
+          // present in the set encoding operation.
+          propEncoding.resultEncodings.push_back(encoding);
+          return propEncoding;
+        })
+        .Default([&](auto) { return failure(); });
+  }
+};
+
+struct ContractionOpPropagationInterface
+    : public IREE::Encoding::EncodingPropagationOpInterface::ExternalModel<
+          ContractionOpPropagationInterface, tensor::CollapseShapeOp> {
+  FailureOr<IREE::Encoding::PropagationResult>
+  propagateEncoding(Operation *op, RewriterBase &builder,
+                    IREE::Encoding::PropagationEncoding encodings,
+                    OpResult opResult) const {
```

Comment:
Same issue for the implementation. Please don't use blank lines when you don't have to. We don't put a blank line between every statement.

https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

> This is more a principle than a rule: don't use blank lines when you don't have to. In particular, don't put more than one or two blank lines between functions, resist starting functions with a blank line, don't end functions with a blank line, and be sparing with your use of blank lines. A blank line within a block of code serves like a paragraph break in prose: visually separating two thoughts.